File: Leetcode/Algorithm/java/00003-Longest Substring Without Repeating Characters.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> st = new HashSet<>();
        int p = 0;
        int ans = 0;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            while (st.contains(c)) {
                st.remove(s.charAt(p));
                p++;
            }
            st.add(c);
            ans = Math.max(ans, st.size());
        }
        return ans;
    }
}

File: Leetcode/Algorithm/kotlin/00003-Longest Substring Without Repeating Characters.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        var st = HashSet<Char>()
        val n = s.length
        
        var l = 0
        var r = 0
        var res = 0
        
        while (l < n && r < n) {
            val cur = s[r]
            while (st.contains(cur)) {
                st.remove(s[l])
                l++
            }
            st.add(s[r])
            r++;
            res = maxOf(res, r - l)
        }
        return res
    }
}
